Implement NFT Class reveal queue abstract

hochiw opened the pull request:
## Features
- Scaffolded ClassRevealQueue map via ignite cli
- Created methods to interact with the queue 
- Created an abstract abci endblocker to process the queue at the end of the block, for use after oursky/likecoin-chain#206 
- Enqueue class for reveal on blindbox creation
- Update or enqueue class for reveal on blindbox update

## Testcases
- Created testcase for inserting entries into the queue
- Created testcase for deleting entry from the queue
- Created testcase for updating an entry from the queue 
- Created testcase to check the sorting of the queue
- Created testcase scaffold for the abci, for use after oursky/likecoin-chain#206 

refs oursky/likecoin-chain#205

hochiw committed:
Add class reveal event proto

refs oursky/likecoin-chain#205

hochiw committed:
Create abci end blocker to process and iterate through the reveal queue

refs oursky/likecoin-chain#205

hochiw committed:
Scaffold classRevealQueue

ignite scaffold map classRevealQueue --index revealTime,classId --module likenft --no-message

then replace revealTime to timestamp

refs oursky/likecoin-chain#205

hochiw committed:
Remove queries for ClassRevealQueues

refs oursky/likecoin-chain#205

hochiw committed:
Change revealTime to timestamp type and implement queue functions

refs oursky/likecoin-chain#205

hochiw committed:
Create abstract testcase for abci

refs oursky/likecoin-chain#205

chihimng commented on proto/likenft/class_reveal_queue.proto:
```diff
@@ -3,9 +3,11 @@ package likecoin.likechain.likenft;
+
 message ClassRevealQueue {
-  string revealTime = 1; 
+  google.protobuf.Timestamp revealTime = 1 [(gogoproto.stdtime) = true]; 
```
set nullable to false as well? so the generated go type will not be pointer

chihimng commented on proto/likenft/event.proto:
```diff
@@ -15,6 +15,10 @@ message EventUpdateClass {
 
+message EventRevealClass {
+  string classId = 1; 
+}
```
can we add class parent fields here for easier listening?

chihimng commented on x/likenft/abci.go:
```diff
@@ -0,0 +1,38 @@
+
+		// TODO: Remove class from queue #206
+
+		ctx.EventManager().EmitTypedEvent(&types.EventRevealClass{
```
this should be part of business logic, not queue level, but let's leave it here for now

chihimng replied:
add a todo for refactoring maybe?

chihimng commented on x/likenft/abci.go:
```diff
@@ -0,0 +1,38 @@
+		ctx.EventManager().EmitTypedEvent(&types.EventRevealClass{
+			ClassId: classId,
+		})
+
```
btw, can we see how to handle exceptions? like where to log & make sure errors will only cause the entry to be skipped and not revert the whole batch?

on exceptions, how to let user trigger retry?

hochiw replied:
errors should be logged by the logger and the entry shall be retried when the next block arrives if it's not removed from the queue

chihimng replied:
- let's add success status & error message at the reveal event
- we should dequeue if failed, and consider if we need to expose a manual reveal method (depending on the failure modes, after we implement the logic)
- i suppose the logic will return a error, but can we capture all unhandled panic / exceptions here?

elise-ng committed:
Rename Queue item to QueueEntry, QueueList to Queue

Ref oursky/likecoin-chain#205

chihimng commented:
Ref issue #205

chihimng commented:
@hochiw you missed `Update queued entry on change` ? on the issue checklist

chihimng commented:
> 

    let's add success status & error message at the reveal event
    we should dequeue if failed, and consider if we need to expose a manual reveal method (depending on the failure modes, after we implement the logic)
    i suppose the logic will return a error, but can we capture all unhandled panic / exceptions here?


I have implemented these just now

hochiw committed:
Remove unused import for query proto

refs oursky/likecoin-chain#205

hochiw committed:
Handle enqueue on new class create and update entry on class update

refs oursky/likecoin-chain#205

hochiw commented:
> @hochiw you missed `Update queued entry on change` ? on the issue checklist

mb you are right, added commits for this

elise-ng committed:
Refactor BlindBoxConfig; Only show toBeRevealed in nft data

Ref oursky/likecoin-chain#205

- to replace feature flag and nullable configs
- use toBeReveal instead of isRevealed flag so the default zero value is more sensible

elise-ng committed:
Update BlindBoxConfig usages; Refactor mint period sorting to sanitize func

Ref oursky/likecoin-chain#205

elise-ng committed:
Fix test cases with BlindBoxConfig usages

Ref oursky/likecoin-chain#205

- Remove case for empty reveal time, as it is no longer nullable

elise-ng committed:
Refactor BlindBoxConfig null check to IsBlindBox()

Ref oursky/likecoin-chain#205

elise-ng committed:
Simplify update logic for reveal queue schedule

Ref oursky/likecoin-chain#205

elise-ng committed:
Do not error if deleting non-existent queue entry

Ref oursky/likecoin-chain#205

elise-ng committed:
Add reveal status in class data; Fix update class overwrites some class data fields

Ref oursky/likecoin-chain#205

elise-ng committed:
Adopt class toBeRevealed flag in mint nft logic

Ref oursky/likecoin-chain#205

elise-ng committed:
Add check for mintable template supply at mint

Ref oursky/likecoin-chain#205

chihimng commented:
@hochiw

Refactored the protos and some logic, also added check for mintable count, can you help cross check?

hochiw commented on x/likenft/keeper/msg_server_update_class.go:
```diff
@@ -82,18 +82,17 @@ func (k msgServer) UpdateClass(goCtx context.Context, msg *types.MsgUpdateClass)
 		k.SetClassRevealQueueEntry(ctx, types.ClassRevealQueueEntry{
 			ClassId:    class.Id,
 			RevealTime: updatedConfig.BlindBoxConfig.RevealTime,
 		})
```
so the update function is not used anymore? but i guess we can keep it just in case

hochiw commented:
@chihimng seems okay

elise-ng committed:
Fix default id and content for unrevealed nft

Ref oursky/likecoin-chain#205

Should have empty url; simplify generated id

chihimng commented:
ran test locally, passes; handle ci timeout later
